File: pysrc/lion_ffi/ffi/_params.py
CTYPEDEF = """
#define LION_FUZZY_SETS_COUNT 8
#define LION_FUZZY_SETS_DEGREE 4
#define LION_SOH_TABLE_COUNT   11

typedef struct lion_params_init {
  double soc;
  double temp_in;
  double soh;
  double capacity;
  double current_guess;
} lion_params_init_t;

typedef struct lion_params_ehc {
  double a;
  double b;
  double mu;
  double kappa;
  double sigma;
  double l;
} lion_params_ehc_t;

typedef struct lion_params_vft {
  double k1;
  double k2;
  double tref;
} lion_params_vft_t;

typedef struct lion_params_temp {
  double cp;
  double rin;
  double rout;
} lion_params_temp_t;

typedef struct lion_params_ocv {
  double alpha;
  double beta;
  double gamma;
  double v0;
  double vl;
} lion_params_ocv_t;

typedef enum lion_rint_model {
  LION_RINT_MODEL_FIXED,
  LION_RINT_MODEL_POLARIZATION,
} lion_rint_model_t;

typedef struct lion_params_rint_fixed {
  double internal_resistance;
} lion_params_rint_fixed_t;

typedef struct lion_params_rint_polarization {
  lion_mf_sigmoid_params_t  c40;
  lion_mf_gaussian_params_t c20;
  lion_mf_gaussian_params_t c10;
  lion_mf_gaussian_params_t c4;

  lion_mf_gaussian_params_t d5;
  lion_mf_gaussian_params_t d10;
  lion_mf_gaussian_params_t d15;
  lion_mf_sigmoid_params_t  d30;

  double poly[LION_FUZZY_SETS_COUNT][LION_FUZZY_SETS_DEGREE];
} lion_params_rint_polarization_t;

typedef struct lion_params_rint {
  lion_rint_model_t model;
  union {
    lion_params_rint_fixed_t        fixed;
    lion_params_rint_polarization_t polarization;
  } params;
} lion_params_rint_t;

typedef struct lion_params_degradation_element {
  double max;
  double min;
  double coeff;
} lion_params_degradation_element_t;

typedef struct lion_params_soh_vendor {
  uint64_t total_cycles;
  double   final_soh;
} lion_params_soh_vendor_t;

typedef struct lion_params_soh_masserano {
  uint64_t                          total_cycles;
  double                            final_soh;
  lion_params_degradation_element_t table[LION_SOH_TABLE_COUNT];
} lion_params_soh_masserano_t;

typedef struct lion_params_soh {
  lion_soh_model_t model;
  union {
    lion_params_soh_vendor_t    vendor;
    lion_params_soh_masserano_t masserano;
  } params;
} lion_params_soh_t;

typedef struct lion_params {
  lion_params_init_t init;
  lion_params_ehc_t ehc;
  lion_params_ocv_t ocv;
  lion_params_vft_t vft;
  lion_params_temp_t temp;
  lion_params_rint_t rint;
} lion_params_t;
"""


CDEF = """
lion_params_init_t              lion_params_default_init(void);
lion_params_ehc_t               lion_params_default_ehc(void);
lion_params_ocv_t               lion_params_default_ocv(void);
lion_params_vft_t               lion_params_default_vft(void);
lion_params_temp_t              lion_params_default_temp(void);
lion_params_rint_fixed_t        lion_params_default_rint_fixed(void);
lion_params_rint_polarization_t lion_params_default_rint_polarization(void);
lion_params_rint_t              lion_params_default_rint(void);
lion_params_soh_t               lion_params_default_soh(void);
lion_params_t                   lion_params_default(void);
"""

File: src/lion_app/params.c
#include "lion/params.h"

#include <lion/lion.h>

#define LION_PARAMS_DEFAULT_INIT                                                                                                                     \
  {                                                                                                                                                  \
    .soc           = 0.1,                                                                                                                            \
    .temp_in       = 298.0,                                                                                                                          \
    .soh           = 1.0,                                                                                                                            \
    .capacity      = 14400.0,                                                                                                                        \
    .current_guess = 10.0,                                                                                                                           \
  }

#define LION_PARAMS_DEFAULT_EHC                                                                                                                      \
  {                                                                                                                                                  \
    .a     = 4e-5,                                                                                                                                   \
    .b     = 5e-5,                                                                                                                                   \
    .mu    = 0.4,                                                                                                                                    \
    .kappa = 3,                                                                                                                                      \
    .sigma = 0.05,                                                                                                                                   \
    .l     = 7,                                                                                                                                      \
  }

#define LION_PARAMS_DEFAULT_OCV                                                                                                                      \
  {                                                                                                                                                  \
    .alpha = 0.15,                                                                                                                                   \
    .beta  = 17,                                                                                                                                     \
    .gamma = 10.5,                                                                                                                                   \
    .v0    = 4.14,                                                                                                                                   \
    .vl    = 3.977,                                                                                                                                  \
  }

#define LION_PARAMS_DEFAULT_VFT                                                                                                                      \
  {                                                                                                                                                  \
    .k1   = -5.738,                                                                                                                                  \
    .k2   = 209.9,                                                                                                                                   \
    .tref = 298,                                                                                                                                     \
  }

#define LION_PARAMS_DEFAULT_TEMP                                                                                                                     \
  {                                                                                                                                                  \
    .cp   = 100.0,                                                                                                                                   \
    .rin  = 3.0,                                                                                                                                     \
    .rout = 9.0,                                                                                                                                     \
  }

#define LION_PARAMS_DEFAULT_RINT_FIXED {.internal_resistance = 0.12}

#define LION_PARAMS_DEFAULT_RINT_POLARIZATION                                                                                                        \
  {                                                                                                                                                  \
    .c40 =                                                                                                                                           \
        {                                                                                                                                            \
              .a = -19.9748,                                                                                                                             \
              .c = -26.5422,                                                                                                                             \
              },                                                                                                                                           \
    .c20 =                                                                                                                                           \
        {                                                                                                                                            \
              .mean  = -20.0,                                                                                                                            \
              .sigma = 3.0,                                                                                                                              \
              },                                                                                                                                           \
    .c10 =                                                                                                                                           \
        {                                                                                                                                            \
              .mean  = -10.0,                                                                                                                            \
              .sigma = 2.3875,                                                                                                                           \
              },                                                                                                                                           \
    .c4 =                                                                                                                                            \
        {                                                                                                                                            \
              .mean  = -4.0,                                                                                                                             \
              .sigma = 2.1623,                                                                                                                           \
              },                                                                                                                                           \
    .d5 =                                                                                                                                            \
        {                                                                                                                                            \
              .mean  = 5.0,                                                                                                                              \
              .sigma = 2.0,                                                                                                                              \
              },                                                                                                                                           \
    .d10 =                                                                                                                                           \
        {                                                                                                                                            \
              .mean  = 10.0,                                                                                                                             \
              .sigma = 3.1631,                                                                                                                           \
              },                                                                                                                                           \
    .d15 =                                                                                                                                           \
        {                                                                                                                                            \
              .mean  = 15.0,                                                                                                                             \
              .sigma = 2.0,                                                                                                                              \
              },                                                                                                                                           \
    .d30 =                                                                                                                                           \
        {                                                                                                                                            \
              .a = 15.9494,                                                                                                                              \
              .c = 17.3438,                                                                                                                              \
              },                                                                                                                                           \
    .poly =                                                                                                                                          \
        {                                                                                                                                            \
              {0.04172, 0.001688, -0.01526, 0.04006},                                                                                                    \
              {0.04385, 0.01758, -0.04159, 0.05488},                                                                                                     \
              {0.05166, 0.02408, -0.05132, 0.06101},                                                                                                     \
              {0.07004, 0.03910, -0.05345, 0.05015},                                                                                                     \
              {0.1317, -0.05083, -0.2579, 0.3084},                                                                                                       \
              {0.0958, -0.05706, -0.07709, 0.1141},                                                                                                      \
              {0.07868, -0.05782, -0.008633, 0.04612},                                                                                                   \
              {0.07218, -0.07066, 0.04202, 0.0061},                                                                                                      \
              },                                                                                                                                           \
}

#define LION_PARAMS_DEFAULT_RINT                                                                                                                     \
  {                                                                                                                                                  \
    .model        = LION_RINT_MODEL_FIXED,                                                                                                           \
    .params.fixed = LION_PARAMS_DEFAULT_RINT_FIXED,                                                                                                  \
  }

#define LION_PARAMS_DEFAULT                                                                                                                          \
  {                                                                                                                                                  \
    .init = LION_PARAMS_DEFAULT_INIT,                                                                                                                \
    .ehc  = LION_PARAMS_DEFAULT_EHC,                                                                                                                 \
    .ocv  = LION_PARAMS_DEFAULT_OCV,                                                                                                                 \
    .vft  = LION_PARAMS_DEFAULT_VFT,                                                                                                                 \
    .temp = LION_PARAMS_DEFAULT_TEMP,                                                                                                                \
    .rint = LION_PARAMS_DEFAULT_RINT,                                                                                                                \
  }

lion_params_init_t lion_params_default_init(void) {
  lion_params_init_t out = LION_PARAMS_DEFAULT_INIT;
  return out;
}
lion_params_ehc_t lion_params_default_ehc(void) {
  lion_params_ehc_t out = LION_PARAMS_DEFAULT_EHC;
  return out;
}
lion_params_ocv_t lion_params_default_ocv(void) {
  lion_params_ocv_t out = LION_PARAMS_DEFAULT_OCV;
  return out;
}
lion_params_vft_t lion_params_default_vft(void) {
  lion_params_vft_t out = LION_PARAMS_DEFAULT_VFT;
  return out;
}
lion_params_temp_t lion_params_default_temp(void) {
  lion_params_temp_t out = LION_PARAMS_DEFAULT_TEMP;
  return out;
}
lion_params_rint_fixed_t lion_params_default_rint_fixed(void) {
  lion_params_rint_fixed_t out = LION_PARAMS_DEFAULT_RINT_FIXED;
  return out;
}
lion_params_rint_polarization_t lion_params_default_rint_polarization(void) {
  lion_params_rint_polarization_t out = LION_PARAMS_DEFAULT_RINT_POLARIZATION;
  return out;
}
lion_params_rint_t lion_params_default_rint(void) {
  lion_params_rint_t out = LION_PARAMS_DEFAULT_RINT;
  return out;
}
lion_params_t lion_params_default(void) {
  lion_params_t out = LION_PARAMS_DEFAULT;
  return out;
}
